syncer/pipe.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"sync"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/gtid"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/errors"
+	"github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+	"github.com/siddontang/go/sync2"
+)
+
+// Pipe is the littlest process unit in syncer
+type Pipe interface {
+	// Name is the pipe's name
+	Name() string
+
+	// Init initializes the pipe
+	Init(*config.SubTaskConfig, func() error) error
+
+	// Update updates the pipe's config
+	Update(*config.SubTaskConfig)
+
+	// Input receives data
+	Input() chan *PipeData
+
+	// Output exports data already be processed
+	Output() chan *PipeData
```

Comment:
the purpose of defining these two methods (`Input` and `Output`) is to make running pipes concurrently someday?

Reply:
there is a `link` function used to link two pipe, get data from one pipe's output, and then send data to another pipe's input

Reply:
what is the advantage of using a `chan` rather than just passing/returning as values? speed up the processing?

Reply:
every pipe will process data async, this will enhance the performance

Reply:
now, one pipe is always bound to a special previous pipe and a special next pipe, right?

Reply:
yes, do you have any suggestions?

Reply:
maybe extract data channel from Pipe? one or more Pipe send data into the channel and one or more Pipe receives data from the channel, but we need to keep the order in some cases, this may be not easy. you can keep considering this.